### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by KarstenMalle**
@prasadtalasila 
Could you please try to check this prior to our bachelor deadline tomorrow.
Hopefully this PR is done correct, it should be rebased with upstream and not merged. 
For more information check #70.

### CODE REVIEW

**Comment by prasadtalasila** [comment]
@KarstenMalle , there is a merge conflict. I am not sure why though

**Comment by Artin13** [comment]
You had rebased the authentication branch with feature? Seems like if do a pr to feature there will not be merge conflicts

**Comment by prasadtalasila** [comment]
> @prasadtalasila did you rebase the authentication branch with feature? Seems like if do a pr to feature there will not be merge conflicts

Yes, that's what @KarstenMalle wanted yesterday.

**Comment by Artin13** [comment]
> > @prasadtalasila did you rebase the authentication branch with feature? Seems like if do a pr to feature there will not be merge conflicts
> 
> Yes, that's what @KarstenMalle wanted yesterday.

@prasadtalasila should we create a pr for feature instead then?

**Comment by KarstenMalle** [comment]
> @KarstenMalle , there is a merge conflict. I am not sure why though

The merge conflict were due to the changes made to the code to fix e2e testing. The conflicts have now been resolved.

**Comment by prasadtalasila** [comment]
I want to see the updates over the committed authenticstion code so that the review can be done quickly for you. If you rebase to authentication, that's quickest. Can you do the rebase to authentication quickly? 

I am already going through the whole code including the one committed to the authentication branch but having only the changed code helps a great deal.

**Comment by prasadtalasila** [comment]
Here is a quick suggestion, if you like it.

1. Rebase to authentication branch
2. Squash all the PR commits to just one.

**Comment by KarstenMalle** [comment]
It should be fine to view the changes here [https://github.com/INTO-CPS-Association/DTaaS/pull/71/files](https://github.com/INTO-CPS-Association/DTaaS/pull/71/files) right?

File: client/test/unitTests/testUtils.tsx [hunk 1]
```diff
@@ -16,7 +16,10 @@ type RouterOptions = {
   store?: Store;
 };
 
-export const renderWithRouter = (ui: React.ReactElement, { route = '/', store }: RouterOptions = {}) => {
+export const renderWithRouter = (
```
Comment (prasadtalasila) [inline client/test/unitTests/testUtils.tsx:19]: Is this for readability?

File: client/test/unitTests/testUtils.tsx [hunk 2]
```diff
@@ -25,9 +28,7 @@ export const renderWithRouter = (ui: React.ReactElement, { route = '/', store }:
           <RouterComponent ui={ui} route={route} />
         </Provider>
       )
-    : render(
-        <RouterComponent ui={ui} route={route} />
-      );
+    : render(<RouterComponent ui={ui} route={route} />);
```
Comment (prasadtalasila) [inline client/test/unitTests/testUtils.tsx:31]: Is this change only for readability?

File: client/test/e2e/playwright/auth.setup.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+const authFile = 'playwright/.auth/user.json';
+
+setup('authenticate', async ({ page }) => {
+  // Perform authentication steps for authentication process.
+  await page.goto('./');
+  await page
+    .getByRole('button', { name: 'GitLab logo Sign In with GitLab' })
+    .click();
+  await page.getByRole('button', { name: 'Accept All Cookies' }).click();
+  await page.fill('#user_login', 'TestUserDTaaS'); // Insert valid GitLab testing username.
```
Comment (prasadtalasila) [inline client/test/e2e/playwright/auth.setup.ts:12]: please bring these out to env file of test/ directory. You can make it into .env and load it via [dotnev](https://github.com/motdotla/dotenv) package.

**Review by prasadtalasila — COMMENTED** [review]
There is no need to change code in `servers/` path.

File: client/test/unitTests/testUtils.tsx (hunk 1 above)
Comment (KarstenMalle) [reply client/test/unitTests/testUtils.tsx:19]: Yes this is formatting

File: client/test/unitTests/testUtils.tsx (hunk 2 above)
Comment (KarstenMalle) [reply client/test/unitTests/testUtils.tsx:31]: This is formatting

File: client/test/unitTests/Components/PrivateRoute.test.tsx [hunk 4]
```diff
@@ -30,7 +30,8 @@ const setupTest = (authState: AuthState) => {
     <PrivateRoute>
       <TestComponent />
     </PrivateRoute>,
-    { route: '/private' },
+
+    { route: '/private' }
```
[the remark this answers is not in the document]
Comment (KarstenMalle) [reply client/test/unitTests/Components/PrivateRoute.test.tsx:34]: It was giving conflicts so I just chose one, it does the same thing so dosent matter.

**Comment by KarstenMalle** [comment]
> There is no need to change code in `servers/` path.

Yes we know, we haven't touched servers/ this is just due to the rebase adding Phillips new PR

**Comment by KarstenMalle** [comment]
> This file is already in upstream `authentication` branch, but what is the difference between jest.config.json and jest.integration.config.json?
> 
> The following lines were removed from jest.integration.config.json.
> 
> ```
>   "coverageThreshold": {
>     "global": {
>       "branches": 20,
>       "functions": 30,
>       "lines": 50,
>       "statements": 50
>     }
>   },
> ```

I did want to set coverageThreshold that would execute the integration tests with exit code 1 due to the threshold not being met by the integration tests since we only have 3 tests. This is anywhere close to hitting those thresholds that we set for the unit tests so therefore I just removed it so the integration tests works.

**Comment by KarstenMalle** [comment]
> A shell script error in `client/script/test.bash`:
> 
> ```
> + export PATH                                                                              
> + '[' -z '' ']'                                                                            
> + mode=unit-tests                                                                          
> + '[' unit-tests == -a ']'                                                                 
> + '[' unit-tests == -u ']' or '[' unit-tests == -unit-tests ']'                            
> script/test.bash: line 27: [: too many arguments                                           
> + '[' unit-tests == -i ']'                                                                 
> + '[' unit-tests == -e ']'                                                                 
> + printf 'Running unit tests only...\n'                                                    
> ```

Fixed

**Comment by Artin13** [comment]
Is it possible to merge into authentication because it is fixed. Just need the minor updates and the test works? @prasadtalasila

**Comment by Artin13** [comment]
If it is possible we can remove the error regarding e2e from our report

**Comment by KarstenMalle** [comment]
@prasadtalasila These are minor changes, can't you just merge so we can hand in the report as we have written it. It should all work and I will push these minor changes very shortly.

**Comment by KarstenMalle** [comment]
Thank you Prasad! :)

**Comment by prasadtalasila** [comment]
> Thank you Prasad! :)

welcome

**Pull request merged**